chainerx_cc/chainerx/cuda/cuda_device/linalg.cu

```diff
@@ -0,0 +1,208 @@
+    throw DtypeError{"Only Arrays of float or double type are supported by getrf (LU)"};
+}
+
+template <typename T>
+cusolverStatus_t Getrf(
+        cusolverDnHandle_t /*handle*/, int /*m*/, int /*n*/, T* /*a*/, int /*lda*/, T* /*workspace*/, int* /*devipiv*/, int* /*devinfo*/) {
+    throw DtypeError{"Only Arrays of float or double type are supported by getrf (LU)"};
+}
+
+template <typename T>
+cusolverStatus_t Getrs(
+        cusolverDnHandle_t /*handle*/,
+        cublasOperation_t /*trans*/,
+        int /*n*/,
+        int /*nrhs*/,
+        T* /*a*/,
+        int /*lda*/,
+        int* /*devipiv*/,
+        T* /*b*/,
+        int /*ldb*/,
+        int* /*devinfo*/) {
+    throw DtypeError{"Only Arrays of float or double type are supported by getrs (Solve)"};
+}
+
+template <>
+cusolverStatus_t GetrfBuffersize<double>(cusolverDnHandle_t handle, int m, int n, double* a, int lda, int* lwork) {
+    return cusolverDnDgetrf_bufferSize(handle, m, n, a, lda, lwork);
+}
+
+template <>
+cusolverStatus_t GetrfBuffersize<float>(cusolverDnHandle_t handle, int m, int n, float* a, int lda, int* lwork) {
+    return cusolverDnSgetrf_bufferSize(handle, m, n, a, lda, lwork);
+}
+
+template <>
+cusolverStatus_t Getrf<double>(cusolverDnHandle_t handle, int m, int n, double* a, int lda, double* workspace, int* devipiv, int* devinfo) {
+    return cusolverDnDgetrf(handle, m, n, a, lda, workspace, devipiv, devinfo);
+}
+
+template <>
+cusolverStatus_t Getrf<float>(cusolverDnHandle_t handle, int m, int n, float* a, int lda, float* workspace, int* devipiv, int* devinfo) {
+    return cusolverDnSgetrf(handle, m, n, a, lda, workspace, devipiv, devinfo);
+}
+
+template <>
+cusolverStatus_t Getrs<double>(
+        cusolverDnHandle_t handle,
+        cublasOperation_t trans,
+        int n,
+        int nrhs,
+        double* a,
+        int lda,
+        int* devipiv,
+        double* b,
+        int ldb,
+        int* devinfo) {
+    return cusolverDnDgetrs(handle, trans, n, nrhs, a, lda, devipiv, b, ldb, devinfo);
+}
+
+template <>
+cusolverStatus_t Getrs<float>(
+        cusolverDnHandle_t handle,
+        cublasOperation_t trans,
+        int n,
+        int nrhs,
+        float* a,
+        int lda,
+        int* devipiv,
+        float* b,
+        int ldb,
+        int* devinfo) {
+    return cusolverDnSgetrs(handle, trans, n, nrhs, a, lda, devipiv, b, ldb, devinfo);
+}
+
+template <typename T>
+void SolveImpl(const Array& a, const Array& b, const Array& out) {
+    Device& device = a.device();
+    Dtype dtype = a.dtype();
+
+    cuda_internal::DeviceInternals& device_internals = cuda_internal::GetDeviceInternals(static_cast<CudaDevice&>(device));
+
+    Array lu_matrix = Empty(a.shape(), dtype, device);
+    device.backend().CallKernel<CopyKernel>(a.Transpose(), lu_matrix);
+    T* lu_ptr = static_cast<T*>(internal::GetRawOffsetData(lu_matrix));
+
+    int m = a.shape()[0];
+    int nrhs = 1;
+    if (b.ndim() == 2) {
+        nrhs = b.shape()[1];
```

Comment:
We generally try to be explicit with int widths, e.g. int64_t for Shape/Dims elements https://github.com/chainer/chainer/blob/master/chainerx_cc/chainerx/dims.h#L11.